### ISSUE TITLE
callbacks

### ISSUE THREAD

**Issue by dustinvtran** [blei-lab/edward#106]
Currently we print the iteration and the variational objective evaluated at that iteration. We'd like to extend to a number of options. Each option is a function that is applied at particular stages of the inference procedure, which can be used to view intermediate states of the model and algorithm during training. 

Some callback examples: loss on test data, logs for TensorBoard, saving the model at intermediate training stages. Most generally, the API should let the user define a custom function as the callback.

**Comment by bayerj** [comment]
We pursued a different path for climin (http://climin.readthedocs.org), a neural net aware  optimisation library. We basically turned the "optimise" call into an iteration instead, yielding info dictionaries containing optimiser specific information. 

This made it possible to give control back to the user after each step, and feels much more natural than callbacks. further, different functionalities share the same scope, so that certain variables can be reused. also, it is much easier to debug.

maybe this approach is also the right one for edward.

**Comment by dustinvtran** [comment]
There's been some discussion with Google Brain collaborators on what things inference should automate and what things should be done manually (enabling flexibility for the user). What we settled upon was allowing both options: (1) a user can call `inference.run()`, which handles everything; or (2) a user can perform manual updates and do additional things during each iteration (e.g., in this [bayesian linear regression](https://github.com/blei-lab/edward/blob/master/examples/bayesian_linear_regression_placeholder_tensorboard.py) script).

I'm very much in favor of your suggestion. At the moment, the manual approach only outputs the loss for debugging. In order to realize this even further, we should have some implementation that outputs an info dictionary for the update, where the info dictionary might contain other optimizer-specific info.

One question would be whether callbacks should still be allowed if the user does not aim to manually run updates. My guess would be no, because it's a weird scenario where you want _some_ control over inference but not all of it.
